llvm_mode/compare-transform-pass.so.cc

```diff
@@ -389,99 +389,133 @@ bool CompareTransform::transformCmps(Module &M, const bool processStrcmp,
 
     }
 
+    if (isConstSized) {
+
+      constSizedLen = dyn_cast<ConstantInt>(sizedValue)->getZExtValue();
+
+    }
+
     if (HasStr1) {
 
       TmpConstStr = Str1.str();
       VarStr = Str2P;
-      constLen = isMemcmp ? sizedLen : TmpConstStr.length();
 
     } else {
 
       TmpConstStr = Str2.str();
       VarStr = Str1P;
-      constLen = isMemcmp ? sizedLen : TmpConstStr.length();
 
     }
 
-    /* properly handle zero terminated C strings by adding the terminating 0 to
-     * the StringRef (in comparison to std::string a StringRef has built-in
-     * runtime bounds checking, which makes debugging easier) */
+    // add null termination character implicit in c strings
     TmpConstStr.append("\0", 1);
```

Comment:
this is not your change ... though thinking about it - for memcmp we should not add a \0 maybe?

Reply:
A string literal in c will have a null termination character added to it whether it then goes on to be used in a memcmp comparison or not. And it doesn't make the unrolling any longer, because the `unrollLen` of a memcmp is governed by `n` anyway.

This is a perfectly valid thing to do:

```c
memcmp(bar, "foo", 4);
```

because the 4th char of `foo` can be expected to be a null.

Reply:
yes. but memcmp can also compare non-strings, e.g. memcmp(bar "\x01\x02\x03\x04\x05", 5)
so here a null should only be added if strlen() of the string is one less than the constlen.

Reply:
You should find that in that case the `unrollLen` does indeed get calculated to be 5 unless I've got my logic wrong.

```c
unrollLen = constSizedLen < constStrLen ? constSizedLen : constStrLen;
```

Adding the null here doesn't actually make any difference to it in the end, I just think it's clearer to add it anyway